Enable forwarded for prefix header configuration (#10266)
quarkusio/quarkus: docs/src/main/asciidoc/vertx.adoc

Assuming `/tmp/` is writeable this can be fixed by setting the `vertx.cacheDirBase` property to point to a directory in `/tmp/` for instance in OpenShift by creating an environment variable `JAVA_OPTS` with the value `-Dvertx.cacheDirBase=/tmp/vertx`.

== Running behind a reverse proxy

> Quarkus could be accessed through proxies that additionally generates headers (e.g. `X-Forwarded-Host`) to keep

Reviewer: ```suggestion
Quarkus could be accessed through proxies that additionally generate headers (e.g. `X-Forwarded-Host`) to keep
```

Author: done

information from the client-facing side of the proxy servers that is altered or lost when they are involved.

> In those scenarios, Quarkus can be configured to automatically updates information like protocol, host, port and URI

Reviewer: ```suggestion
In those scenarios, Quarkus can be configured to automatically update information like protocol, host, port and URI
```

Author: done

reflecting the values in these headers.

> IMPORTANT: Activate this feature makes the server exposed to several security issues (i.e. information spoofing).

Reviewer: ```suggestion
IMPORTANT: Activating this feature makes the server exposed to several security issues (i.e. information spoofing).
```

Consider activate it only when running behind a reverse proxy.

To setup this feature, please include the following lines in `src/main/resources/application.properties`:
[source]
----
quarkus.http.proxy-address-forwarding=true
----

To consider only de-facto standard header (`Forwarded` header), please include the following lines in `src/main/resources/application.properties`:
[source]
----
quarkus.http.allow-forwarded=true
----

To consider only non-standard headers, please include the following lines instead in `src/main/resources/application.properties`:

[source]
----
quarkus.http.proxy-address-forwarding=true
quarkus.http.proxy.enable-forwarded-host=true
quarkus.http.proxy.enable-forwarded-prefix=true
----

> Both configurations related with standard and non-standard headers can be combine. Therefore, the

Reviewer: ```suggestion
Both configurations related with standard and non-standard headers can be combined.
```

> `Forwarded` header will have precedence in presence of both set of headers.

Reviewer: ```suggestion
In this case, the `Forwarded` header will have precedence in presence of both set of headers.
```

Author: done

> Supported forwarding address headers:

Reviewer: ```suggestion
Supported forwarding address headers are:
```

Author: done

* `Forwarded`
* `X-Forwarded-Proto`
* `X-Forwarded-Host`
* `X-Forwarded-Port`

> * `X-Forwarded-Prefix`

Reviewer: I'm not sure I'm following the purpose of this list?

Also are the two headers config documented somewhere?

Author: My intention was to provide a list of headers that Quarkus can use to rewrite request's data. As I didn't find any documentation that expose the list, I decided to include it. 

If you think this list shouldn't be here, I can remove it.

Author: The `X-Forwarded-Prefix` header config support is something new. Some ingress proxies (e.g. nginx) generates this header when they remove a prefix in the request's path to keep the original info. However, you have special cases like envoy which it use another header, x-envoy-path if I am not in mistake, for the same purpose. Hence, I let this header "configurable".

Author: You can check these config properties [here].(https://github.com/quarkusio/quarkus/blob/f00103df7f39b09c0e46563da33749599b52d1ed/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ProxyConfig.java)
